classes/sample_set.php

```diff
@@ -17,31 +17,85 @@
 namespace tool_excimer;
 
 /**
- * Stores samples copied over from the profiler.
+ * Stores samples copied over from the profiler, to be used in a profile.
  *
  * @package    tool_excimer
  * @author     Jason den Dulk <jasondendulk@catalyst-au.net>
  * @copyright  2022 Catalyst IT
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-
 class sample_set {
     public $name;
-    public $samples = [];
     public $starttime;
 
+    public $samples = [];
+
+    public $samplelimit;
+
+    /** @var int If $filterrate is R, then only each Rth sample is recorded. */
+    private $filterrate = 1;
 
-    public function __construct($name, $starttime) {
+    /** @var int Internal counter to help with filtering. */
+    private $counter = 0;
+
+    /**
+     * Constructs the sample set.
+     *
+     * @param string $name
+     * @param float $starttime
+     * @param int $samplelimit
+     */
+    public function __construct(string $name, float $starttime, int $samplelimit) {
         $this->name = $name;
         $this->starttime = $starttime;
+        $this->samplelimit = $samplelimit;
     }
 
     /**
-     * Add a sample the sample store.
+     * Add a sample to the sample store, applying any filters.
      *
      * @param array $sample
      */
-    public function add_sample(\ExcimerLogEntry $sample): void {
-        $this->samples[] = $sample;
+    public function add_sample(\ExcimerLogEntry $sample) {
+        if (count($this->samples) === $this->samplelimit) {
+            $this->apply_doubling();
+        }
+        $this->counter += 1;
+        if ($this->counter === $this->filterrate) {
+            $this->samples[] = $sample;
+            $this->counter = 0;
+        }
+    }
+
+    /**
+     * Add a number of samples.
+     *
+     * @param iterable $samples
+     */
+    public function add_many_samples(iterable $samples) {
+        foreach ($samples as $sample) {
+            $this->add_sample($sample);
+        }
+    }
+
+    /**
+     * Doubles the filter rate, and strips every second sample from the set.
+     * Called when the sample limit is reached.
+     *
+     * We have two options here. Either double the sampling period, or apply a filter to record only the
+     * Nth sample passed to sample_set. By using a filter and keeping the sampling period the same, we avoid
+     * spilling over into the next task.
+     */
+    public function apply_doubling() {
+        $this->filterrate *= 2;
+        $this->samples = array_values(
+            array_filter(
+                $this->samples,
+                function($key) {
```

Comment:
I'm not sure this algorithm is correct. We are removing every second entry in the list of samples but this isn't exactly the same as halving the samples. Consider this test case with 112 time slices:

```
1> a > b > c (1 sample)
2> a > b > c (1 sample)
3> a > b > c (10 samples)
4> a > b > c (100 samples)
```
 if we remove every second record then we could end up with this only 11 time slices 

```
1> a > b > c (1 sample)
3> a > b > c (10 samples)
```

but this isn't what we want, what we really want is this with 56 time slices = 112 * .5:

```
1> a > b > c (1 sample)
3> a > b > c (50 samples)
4> a > b > c (5 samples)
```

Reply:
Also I wonder if `apply_doubling` should be a private method. Aside from it being used in unit tests, do you think this would ever be used outside of the logic when adding a new sample?

Reply:
@brendanheywood From my read it seems like the samples would be added one at a time, so the removal of every 2nd happens as part of the samples being added instead of them being added as a batch, then processed afterwards. But @jaypha feel free to correct me if this is wrong.